File: 8-string-to-integer-atoi/string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        s = s.trim();
        int n = s.length();
        if (n == 0) {
            return 0;
        }
        long ans = 0;
        boolean neg = false;
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (i == 0) {
                if (ch == '-') {
                    neg = true;
                    continue;
                } else if (ch == '+') {
                    neg = false;
                    continue;
                }
            }
            if (Character.isDigit(ch)) {
                int dig = ch - '0';
                ans = ans * 10 + dig;
                if (neg) {
                    long check = -ans;
                    if (check < Integer.MIN_VALUE) {
                        return Integer.MIN_VALUE;
                    }
                } else {
                    if (ans > Integer.MAX_VALUE) {
                        return Integer.MAX_VALUE;
                    }
                }
            } else {
                break;
            }
        }
        if (neg) {
            ans = -ans;
        }
        return (int)ans;
    }
}

File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int n = s.size();
        int i =0;
        int sign = 1;
        int ans =0;
        while(i<n && s[i] == ' '){
            i++;
        }

        if(i<n && (s[i] == '-' || s[i] == '+')){
            sign = (s[i] == '-')?-1:1;
            i++;
        }

        while(i<n && isdigit(s[i])){
            int digit = s[i] - '0';
            if (ans > (INT_MAX - digit) / 10) {
                return (sign == 1) ? INT_MAX : INT_MIN;
            }

            ans = ans * 10 + digit;
            i++;
        }
        return ans*sign;
    }
};